### ISSUE TITLE
converge_if_changed ignore missing required properties

### ISSUE THREAD

**Issue by Annih** [chef/chef#5992]
## Description

The `converge_if_changed` method is only checking "properties specified by the user on the resource" to determine whether the block should be called or not. Meaning that if the user forget to pass my important property, my block protected by a `converge_if_changed` won't be called, and the `requirement` validation will never happen.

I already read #4596, but I think this issue is a more specific case, which I personally qualify as a bug what do you think? :)

## Chef Version

12.13.37 & 12.19.36

## Platform Version

* CentOS 6.8, & 7.2
* Windows Server 2012R2 & 2016

## Replication Case

Create a resource

```ruby
resource_name :my_resource

property :important_attribute, String, required: true

action :try_to_do_something do
  converge_if_changed(:important_attribute) do
    puts 'do something'
  end
end
```
And use it:
```ruby
my_resource 'test'
```

**Cc:** @aboten

**Comment by lamont-granquist** [comment]
yeah so the problem here is that we should only be looking at the properties that the user cares about.  consider implementing a file provider using this api where the `new_resource.mode` is `nil` (the default) but the `current_resource.mode` exists and is 0755 because the file already exists.  we definitely don't want to do the comparison between the current_resource and new_resource on `mode` and decide that its 'different'.

so you can't fix this by enforcing that all the properties are checked, because they're not being checked for a good reason.

you might be able to simply run through all the properties given to converge_if_changed in a void context to get them and force the wave function to collapse first.

**Comment by Annih** [comment]
I'm not sure to understand your proposed solution/workaround. Are you saying I have to perform a validation before my `converge_if_changed(:important_and_required_property)`?

This seems very counter productive/intuitive.
Instinctively I would expect that the `converge_if_changed` method to compare the current value of the property to the provided one.
If the property validation is done in a lazy way, I don't care this is an implementation detail, but I would expect the requirement validation to be performed at this point, because I need the value!
My code is saying _do something on this property_ implying _please verify that it's provided first_.

Am I weird to think like that?

**Comment by lamont-granquist** [comment]
no i'm saying that `converge_if_changed` could be modified to do a get of all the properties first (and throw the values away).

> Instinctively I would expect the converge_if_changed method to compare the current value of the property to the provided one.

right, but that gets the case where the provided value is the default/unset/nil value wrong in many cases.

something more or less like this:

```ruby
resource_name :file_perms

property :path, String, name_property: true
property :mode, Integer
property :owner, String
property :group, String

load_current_value do
  if File.exist?(path)
    mode File.stat(path).mode
  end
  [...]
end

action :doit do
  converge_if_changed(:mode) do
    FileUtils.chmod new_resource.mode, new_resource.path
  end
  [...]
end
```

if you call that like:

```ruby
file_perms "/tmp/foo.xy" do
  owner "root"
end
```

then new_resource.mode will be nil, and you need to bypass the converge_if_changed check entirely -- because it is unset.  it very deliberately is skipping unset properties.

converge_if_changed could make sure that properties that were specifically marked required were not skipped, or it might be able to just go through all the properties and do a 'get' on them to force un-lazying all of them.  but it can't just compare all of them.

**Comment by Djailla** [comment]
@Annih based on the last merged commit, is this bug fixed ?

**Comment by lamont-granquist** [comment]
no, i just made a comment in the PR that since this wasn't fixed the workaround used there had to be done, so this is still an open issue

**Comment by lamont-granquist** [comment]
Okay this bug is very confusingly written, and I think it is straight up a duplicate of #4596.  I think I confused myself before because I didn't read this correctly and thought it had something to do with the other converge_if_changed bug.

There's nothing particularly special about converge_if_changed here, we simply need to force validation on all required properties before calling the action or something like that.

**Comment by lamont-granquist** [comment]
Actually this looks like it is just fixed, might have been done by #6300 

```
    Chef::Exceptions::ValidationFailed
    ----------------------------------
    important_attribute is a required property
```
